File: SRXLite.CS/Services/DocumentService.cs
using System;
using System.Collections.Generic;
using System.Web.Services;
using System.Web.Services.Protocols;
using SRXLite.Classes;
using SRXLite.DataTypes;
using static SRXLite.Modules.Util;
using static SRXLite.Modules.ExceptionHandling;

namespace SRXLite.Web.Services
{
    [WebService(Namespace = "http://tempuri.org/srxlite/documentservice")]
    [WebServiceBinding(ConformsTo = WsiProfiles.BasicProfile1_1)]
    [System.ComponentModel.ToolboxItem(false)]
    public class DocumentService : System.Web.Services.WebService
    {
        public ServiceLogin _login;
        private ServiceUser _user;
        private Document _doc;
        private Group _group;
        private Entity _entity;

        #region Constructor

        /// <summary>
        /// Constructor
        /// </summary>
        public DocumentService()
        {
            _login = new ServiceLogin();
            _user = new ServiceUser();
        }

        #endregion

        #region CreateGroup

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginCreateGroup(string groupName, AsyncCallback callback, object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _group = new Group(_user);
                return _group.BeginCreate(groupName, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginCreateGroup", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public long EndCreateGroup(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                return _group.EndCreate(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndCreateGroup", true);
                throw soapEx;
            }
        }

        #endregion

        #region DeleteDocument

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginDeleteDocument(long docID, AsyncCallback callback, object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _doc = new Document(_user, docID);
                return _doc.BeginDelete(callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginDeleteDocument", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public void EndDeleteDocument(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                _doc.EndDelete(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndDeleteDocument", true);
                throw soapEx;
            }
        }

        #endregion

        #region GetEntityDocumentList

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginGetEntityDocumentList(string entityID, AsyncCallback callback, object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _entity = new Entity(_user, entityID);
                return _entity.BeginGetDocumentList(callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginGetEntityDocumentList", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public List<DocumentData> EndGetEntityDocumentList(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                return _entity.EndGetDocumentList(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndGetEntityDocumentList", true);
                throw soapEx;
            }
        }

        #endregion

        #region GetGroupDocumentList

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginGetGroupDocumentList(long groupID, AsyncCallback callback, object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _group = new Group(_user, groupID);
                return _group.BeginGetDocumentList(callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginGetGroupDocumentList", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public List<DocumentData> EndGetGroupDocumentList(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                return _group.EndGetDocumentList(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndGetGroupDocumentList", true);
                throw soapEx;
            }
        }

        #endregion

        #region GetDocumentUploadUrl

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginGetDocumentUploadUrl(
            string entityName,
            int docTypeID,
            long groupID,
            string stylesheetUrl,
            string entityDisplayText,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _doc = new Document(_user);
                return _doc.BeginGetUploadUrl(entityName, docTypeID, groupID, stylesheetUrl, entityDisplayText, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginGetDocumentUploadUrl", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public string EndGetDocumentUploadUrl(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                return _doc.EndGetUploadUrl(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndGetDocumentUploadUrl", true);
                throw soapEx;
            }
        }

        #endregion

        #region GetDocumentViewerUrl

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginGetDocumentViewerUrl(
            long docID,
            bool isReadOnly,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _doc = new Document(_user, docID);
                return _doc.BeginGetViewerUrl(docID, isReadOnly, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginGetDocumentViewerUrl", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public string EndGetDocumentViewerUrl(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                return _doc.EndGetViewerUrl(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndGetDocumentViewerUrl", true);
                throw soapEx;
            }
        }

        #endregion

        #region MoveGroupDocument

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginMoveGroupDocument(
            long docID,
            long sourceGroupID,
            long targetGroupID,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _group = new Group(_user, sourceGroupID);
                return _group.BeginMoveDocument(docID, targetGroupID, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginMoveGroupDocument", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public void EndMoveGroupDocument(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                _group.EndMoveDocument(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndMoveGroupDocument", true);
                throw soapEx;
            }
        }

        #endregion

        #region UpdateDocumentStatus

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginUpdateDocumentStatus(
            long docID,
            DocumentStatus docStatus,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _doc = new Document(_user, docID);
                return _doc.BeginUpdateStatus(docStatus, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginUpdateDocumentStatus", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public void EndUpdateDocumentStatus(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                _doc.EndUpdateStatus(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndUpdateDocumentStatus", true);
                throw soapEx;
            }
        }

        #endregion

        #region UpdateDocumentKeys

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginUpdateDocumentKeys(
            long docID,
            DocumentKeys docKeys,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _doc = new Document(_user, docID);
                return _doc.BeginUpdateKeys(docKeys, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginUpdateDocumentKeys", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public void EndUpdateDocumentKeys(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                _doc.EndUpdateKeys(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndUpdateDocumentKeys", true);
                throw soapEx;
            }
        }

        #endregion

        #region UploadDocument

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginUploadDocument(
            byte[] fileBytes,
            UploadKeys uploadKeys,
            long groupID,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);

                // Validation
                if (!IsFileExtValid(uploadKeys.FileName))
                    throw new Exception("File type not supported.");
                if (!IsFileSizeValid(fileBytes.Length))
                    throw new Exception("File size exceeded the maximum limit of " + GetFileSizeUploadLimitMB() + " MB.");

                _doc = new Document(_user, 0, groupID);
                uploadKeys.InputType = InputType.WebServiceUpload;

                return _doc.BeginUpload(fileBytes, uploadKeys, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginUploadDocument", true);
                throw soapEx;
            }
        }

        [WebMethod(Description = "")]
        [SoapHeader("_login")]
        public long EndUploadDocument(IAsyncResult result)
        {
            try
            {
                if (_doc.HasErrors)
                {
                    throw new Exception(_doc.ErrorMessage);
                }

                _user.Authenticate(_login);
                _doc.EndUpload(result);
                return _doc.DocID;
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndUploadDocument", true);
                throw soapEx;
            }
        }

        #endregion

        #region CopyGroupDocuments

        [WebMethod]
        [SoapHeader("_login")]
        public IAsyncResult BeginCopyGroupDocuments(
            long oldGroupId,
            long newGroupId,
            long oldDocTypeId,
            long newDocTypeId,
            AsyncCallback callback,
            object stateObject)
        {
            try
            {
                _user.Authenticate(_login);
                _group = new Group(_user);
                return _group.BeginCopyGroupDocuments(oldGroupId, newGroupId, oldDocTypeId, newDocTypeId, callback, stateObject);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "BeginCopyGroupDocuments", true);
                throw soapEx;
            }
        }

        [WebMethod]
        [SoapHeader("_login")]
        public void EndCopyGroupDocuments(IAsyncResult result)
        {
            try
            {
                _user.Authenticate(_login);
                _group.EndCopyGroupDocuments(result);
            }
            catch (UnauthorizedAccessException accessEx)
            {
                LogError(accessEx.ToString());
                throw accessEx;
            }
            catch (Exception ex)
            {
                SoapException soapEx = CreateSoapException(ex.Message, ex.ToString(), _user, "EndCopyGroupDocuments", true);
                throw soapEx;
            }
        }

        #endregion

        #region Dispose

        protected override void Dispose(bool disposing)
        {
            if (_doc != null) _doc.Dispose();
            if (_group != null) _group.Dispose();
            if (_entity != null) _entity.Dispose();
            base.Dispose(disposing);
        }

        #endregion
    }
}

File: SRXLite/Services/DocumentService.vb
Option Strict On

Imports System.Web
Imports System.Web.Services
Imports System.Web.Services.Protocols
Imports SRXLite.Classes
Imports SRXLite.DataTypes
Imports SRXLite.Modules

Namespace Web.Services

    <WebService(Namespace:="http://tempuri.org/srxlite/documentservice")> _
    <WebServiceBinding(ConformsTo:=WsiProfiles.BasicProfile1_1)> _
    <Global.Microsoft.VisualBasic.CompilerServices.DesignerGenerated()> _
    Public Class DocumentService
        Inherits System.Web.Services.WebService

        Public _login As ServiceLogin
        Private _user As ServiceUser
        Private _doc As Document
        Private _group As Group
        Private _entity As Entity

#Region " Constructor "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub New()
            _login = New ServiceLogin()
            _user = New ServiceUser()
        End Sub
#End Region

#Region " CreateGroup "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="groupName"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginCreateGroup( _
    ByVal groupName As String, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                'LogError(_login.UserName & " - " & _login.Password)
                _group = New Group(_user)
                Return _group.BeginCreate(groupName, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginCreateGroup", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function EndCreateGroup(ByVal result As IAsyncResult) As Long
            Try
                _user.Authenticate(_login)
                Return _group.EndCreate(result) 'Return groupID

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndCreateGroup", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " DeleteDocument "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="docID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginDeleteDocument( _
    ByVal docID As Long, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _doc = New Document(_user, docID)
                Return _doc.BeginDelete(callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginDeleteDocument", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Sub EndDeleteDocument(ByVal result As IAsyncResult)
            Try
                _user.Authenticate(_login)
                _doc.EndDelete(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndDeleteDocument", True)
                Throw soapEx
            End Try
        End Sub
#End Region

#Region " GetEntityDocumentList "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="entityID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginGetEntityDocumentList( _
     ByVal entityID As String, _
     ByVal callback As AsyncCallback, _
     ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _entity = New Entity(_user, entityID)
                Return _entity.BeginGetDocumentList(callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginGetEntityDocumentList", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function EndGetEntityDocumentList(ByVal result As IAsyncResult) As List(Of DocumentData)
            Try
                _user.Authenticate(_login)
                Return _entity.EndGetDocumentList(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndGetEntityDocumentList", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " GetGroupDocumentList "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="groupID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginGetGroupDocumentList( _
     ByVal groupID As Long, _
     ByVal callback As AsyncCallback, _
     ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _group = New Group(_user, groupID)
                Return _group.BeginGetDocumentList(callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginGetGroupDocumentList", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function EndGetGroupDocumentList(ByVal result As IAsyncResult) As List(Of DocumentData)
            Try
                _user.Authenticate(_login)
                Return _group.EndGetDocumentList(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndGetGroupDocumentList", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " GetDocumentUploadUrl "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="entityName"></param>
        ''' <param name="docTypeID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginGetDocumentUploadUrl( _
     ByVal entityName As String, _
     ByVal docTypeID As Integer, _
     ByVal groupID As Long, _
     ByVal stylesheetUrl As String, _
     ByVal entityDisplayText As String, _
     ByVal callback As AsyncCallback, _
     ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _doc = New Document(_user)
                Return _doc.BeginGetUploadUrl(entityName, docTypeID, groupID, stylesheetUrl, entityDisplayText, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginGetDocumentUploadUrl", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function EndGetDocumentUploadUrl(ByVal result As IAsyncResult) As String
            Try
                _user.Authenticate(_login)
                Return _doc.EndGetUploadUrl(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndGetDocumentUploadUrl", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " GetDocumentViewerUrl "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="docID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginGetDocumentViewerUrl( _
     ByVal docID As Long, _
     ByVal isReadOnly As Boolean, _
     ByVal callback As AsyncCallback, _
     ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _doc = New Document(_user, docID)
                Return _doc.BeginGetViewerUrl(docID, isReadOnly, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginGetDocumentViewerUrl", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function EndGetDocumentViewerUrl(ByVal result As IAsyncResult) As String
            Try
                _user.Authenticate(_login)
                Return _doc.EndGetViewerUrl(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndGetDocumentViewerUrl", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " MoveGroupDocument "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="docID"></param>
        ''' <param name="sourceGroupID"></param>
        ''' <param name="targetGroupID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginMoveGroupDocument( _
    ByVal docID As Long, _
    ByVal sourceGroupID As Long, _
    ByVal targetGroupID As Long, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _group = New Group(_user, sourceGroupID)
                Return _group.BeginMoveDocument(docID, targetGroupID, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginMoveGroupDocument", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Sub EndMoveGroupDocument(ByVal result As IAsyncResult)
            Try
                _user.Authenticate(_login)
                _group.EndMoveDocument(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndMoveGroupDocument", True)
                Throw soapEx
            End Try
        End Sub
#End Region

#Region " UpdateDocumentStatus "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="docID"></param>
        ''' <param name="docStatus"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginUpdateDocumentStatus( _
    ByVal docID As Long, _
    ByVal docStatus As DocumentStatus, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _doc = New Document(_user, docID)
                Return _doc.BeginUpdateStatus(docStatus, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginUpdateDocumentStatus", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Sub EndUpdateDocumentStatus(ByVal result As IAsyncResult)
            Try
                _user.Authenticate(_login)
                _doc.EndUpdateStatus(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndUpdateDocumentStatus", True)
                Throw soapEx
            End Try
        End Sub
#End Region

#Region " UpdateDocumentKeys "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="docID"></param>
        ''' <param name="docKeys"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginUpdateDocumentKeys( _
    ByVal docID As Long, _
    ByVal docKeys As DocumentKeys, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                _doc = New Document(_user, docID)
                Return _doc.BeginUpdateKeys(docKeys, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginUpdateDocumentKeys", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Sub EndUpdateDocumentKeys(ByVal result As IAsyncResult)
            Try
                _user.Authenticate(_login)
                _doc.EndUpdateKeys(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndUpdateDocumentKeys", True)
                Throw soapEx
            End Try
        End Sub
#End Region

#Region " UploadDocument "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="fileBytes"></param>
        ''' <param name="uploadKeys"></param>
        ''' <param name="groupID"></param>
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginUploadDocument( _
    ByVal fileBytes As Byte(), _
    ByVal uploadKeys As UploadKeys, _
    ByVal groupID As Long, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)

                'Validation
                If Not IsFileExtValid(uploadKeys.FileName) Then Throw New Exception("File type not supported.")
                If Not IsFileSizeValid(fileBytes.Length) Then Throw New Exception("File size exceeded the maximum limit of " & GetFileSizeUploadLimitMB() & " MB.")

                _doc = New Document(_user, 0, groupID)
                uploadKeys.InputType = InputType.WebServiceUpload

                Return _doc.BeginUpload(fileBytes, uploadKeys, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginUploadDocument", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod(Description:="")> _
        <SoapHeader("_login")> _
        Public Function EndUploadDocument(ByVal result As IAsyncResult) As Long
            Try
                If _doc.HasErrors Then
                    Throw New Exception(_doc.ErrorMessage)
                End If

                _user.Authenticate(_login)
                _doc.EndUpload(result)
                Return _doc.DocID

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndUploadDocument", True)
                Throw soapEx
            End Try
        End Function
#End Region

#Region " CopyGroupDocuments "
        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="oldGroupId"></param>
        ''' <param name="newGroupId"></param>
        ''' <param name="oldDocTypeId"></param>
        ''' <param name="newDocTypeId"></param>		
        ''' <param name="callback"></param>
        ''' <param name="stateObject"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Function BeginCopyGroupDocuments( _
    ByVal oldGroupId As Long, _
    ByVal newGroupId As Long, _
    ByVal oldDocTypeId As Long, _
    ByVal newDocTypeId As Long, _
    ByVal callback As AsyncCallback, _
    ByVal stateObject As Object) As IAsyncResult

            Try
                _user.Authenticate(_login)
                '_group = New Group(_user, docID)
                _group = New Group(_user)
                Return _group.BeginCopyGroupDocuments(oldGroupId, newGroupId, oldDocTypeId, newDocTypeId, callback, stateObject)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "BeginCopyGroupDocuments", True)
                Throw soapEx
            End Try
        End Function

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="result"></param>
        ''' <remarks></remarks>
        <WebMethod()> _
        <SoapHeader("_login")> _
        Public Sub EndCopyGroupDocuments(ByVal result As IAsyncResult)
            Try
                _user.Authenticate(_login)
                _group.EndCopyGroupDocuments(result)

            Catch accessEx As UnauthorizedAccessException
                LogError(accessEx.ToString)
                Throw accessEx
            Catch ex As Exception
                Dim soapEx As SoapException = CreateSoapException(ex.Message, ex.ToString, _user, "EndCopyGroupDocuments", True)
                Throw soapEx
            End Try
        End Sub
#End Region

#Region " Dispose "
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If _doc IsNot Nothing Then _doc.Dispose()
            If _group IsNot Nothing Then _group.Dispose()
            If _entity IsNot Nothing Then _entity.Dispose()
            MyBase.Dispose(disposing)
        End Sub
#End Region

    End Class

End Namespace